### ISSUE TITLE
dedicated support for Django

### ISSUE THREAD

**Issue by aaschenbrener** [astral-sh/ty#1018]
### Summary

This is an expansion on the known issue:  No implicit import of submodules (https://github.com/astral-sh/ty/issues/133). I've imported Django models explicitly from their submodules. There is an issue with dynamic model attributes like objects and id that Ty doesn’t recognize.

I’m encountering persistent unresolved-attribute errors when running Ty static type checker on Django models, specifically on standard Django model attributes such as id, demo_name, and objects. These errors occur even though the Django model is properly defined, registered, explicitly imported and confirmed to have those attributes in the Django shell.

Steps to reproduce:

Define a Django model with fields such as id, demo_name in apps/sample/models/example.py.

Confirm the model loads and works correctly inside the Django shell:

```py
from apps.sample.models.example import Demo
from django.db import models

issubclass(Demo, models.Model)  # Returns True
Demo._meta.get_fields()  # Returns expected model fields including 'id'
Demo.objects.all()  # Executes without error
```

Run Ty on a test file that uses this model, e.g. test_ty_model.py:

```py
def test_model_fields(m: Demo):
    reveal_type(m.id)
    reveal_type(m.demo_name)
    reveal_type(m.save)
```

Ty reports errors like:

```
error[unresolved-attribute]: Type `Demo` has no attribute `id`
error[unresolved-attribute]: Type `Demo` has no attribute `objects`
```

I have tried running Ty with explicit environment variables to load Django settings, but the errors persist:

```
PYTHONPATH=. DJANGO_SETTINGS_MODULE=proj.settings.dev ty check test_ty_model.py
```

Expected behavior:
Ty should recognize standard Django model attributes (id, objects, model fields) and when the Django environment is correctly configured.

Actual behavior:
Ty reports unresolved-attribute errors for these standard model attributes, even though Django shell confirms their existence.

Environment:

Ty version: 0.0.1-alpha.18 (d697cc092 2025-08-14)
Python version: 3.13.6
Django version: 5.2.5
OS: macOS 15.6 24G84

Project structure:

```
project-root/
├── manage.py
├── test_ty_model.py             
├── apps/
│   ├── sample/
│   │   ├── models/
│   │   │   └── example.py
├── proj/
│   ├── settings/
│   │   └── base.py
│   │   └──  dev.py
```

### Version

0.0.1-alpha.18 (d697cc092 2025-08-14)

**Comment by carljm** [comment]
Thanks for the report! At the moment we don't have any dedicated support for Django, and due to Django's use of dynamic metaclass modifications, some attributes of Django's models aren't visible to a type-checker unless it implements dedicated Django support.

I'm a bit confused by the mention of using `django = true` in `.ty.toml`, since no such option exists (or has ever existed) in ty; seems like maybe something hallucinated by an LLM?

We would like to add some kind of Django support in future; will keep this issue open to track that.

**Comment by MichaReiser** [comment]
We also don't support `.ty.toml`. The configuration file must be named `ty.toml` and the `src` option is `src.root = ["apps", "proj"]`. Python path is under `environment`, etc. See https://docs.astral.sh/ty/reference/configuration/#environment

**Comment by aaschenbrener** [comment]
Thank you for the update. Yes, I was having trouble with django and difficulty figuring out how ty.toml could help and attempted to  on an LLM to help configure. It seems to have hallucinated a lot. I've removed the ty.toml file completely for now.

**Comment by VaishnavGhenge** [comment]
+1. Django is one of the most popular Python frameworks, and these unresolved-attribute errors on model fields make ty unusable for Django projects without sprinkling # ty: ignore everywhere.
Would love to see either native support for Django's model metaclass patterns or a plugin system for django-stubs integration.

@carljm

**Comment by ronaldorcampos** [comment]
I have a lot of errors of the type

Attribute `id` may be missing on object of type `Unknown | ForeignKey` - accessing id in a FK
Attribute `save` may be missing on object of type `Unknown | ForeignKey` - accessing save in a FK
Element `BigIntegerField` of this union is not assignable to `int` - passing pk/id to a function that takes int 
Attribute `hour` may be missing on object of type `Unknown | DateTimeField` - accessing hour of a datetime field
No argument provided for required parameter `cls` of function `as_manager` - overriding objects manager

**Comment by beaugunderson** [comment]
it does sound like there's a [commitment](https://astral.sh/blog/ty#:~:text=Following%20the%20Beta,and%20Django.) to future Django support, which I am REALLY looking forward to!

<img width="749" height="192" alt="Image" src="https://github.com/user-attachments/assets/d6933f75-702a-4112-9c32-80f18e686633" />

**Comment by ramtinabadi** [comment]
Just tinkering with ty, I had a workaround for Django's model using my old package, [django-hint](https://github.com/Vieolo/django-hint), which seems to work until a first-party support is added to ty.

Here is how you use it:

```python
from typing import cast
from django_hint import StandardModelType

class SomeModel(models.Model, StandardModelType['SomeModel']):
    name = cast('str', models.CharField(max_length=100))
```

Adding the `StandardModelType` to the class and casting each field, you wouldn't need to force ty to ignore the type check, and you'll have access to the query functions.

This is just a workaround until first-party support is added

**Comment by rizlas** [comment]
I'm experiencing the same issue in a similar setup. I'm developing NetBox scripts where I provide the NetBox repository (Django-based) as an extra path for better IDE support and code completion:

```toml
[tool.ty.environment]
extra-paths = ["source/netbox/netbox"]
```

My scripts import NetBox models like:

```python
from ipam.models import IPAddress

try:
    ip = IPAddress.objects.get(address="10.0.0.1")
except IPAddress.DoesNotExist:  # ← unresolved-attribute error
    pass
```

Even with the Django source code available in `extra-paths`, Ty still reports `unresolved-attribute` for `DoesNotExist` and `MultipleObjectsReturned` (and others) on all Django models.

Currently I'm disabling `unresolved-attribute` (I know, not good), but it would be great to have proper Django metaclass support.

**Comment by capitantoto** [comment]
I'd love to migrate my company's projects to `ty`, but we depend quite heavily on Django sadly. If there's any low criticality work that a data scientist with mediocre SWE skills could contribute here, I'm more than welcome to do so!

**Comment by McDinii** [comment]
django-types works after https://github.com/astral-sh/ty/releases/tag/0.0.21 release

**Comment by pablo-silva-acin** [comment]
> django-types works after https://github.com/astral-sh/ty/releases/tag/0.0.21 release

Can you give a little more detail on how to get it working?

**Comment by ronaldorcampos** [comment]
> > django-types works after https://github.com/astral-sh/ty/releases/tag/0.0.21 release
> 
> Can you give a little more detail on how to get it working?

you just need to have django-stubs installed.

**Comment by WhyNotHugo** [comment]
django-types seems to necessitate manual type declaration of all the fields implicitly created by Django. A lot of redundant work which is easy to get out-of sync.

django-stubs relies on the mypy plug-in, no?

I tried both, but lots of things fail, like the reverse side of a foreign key fails with `unresolved-attribute`.

**Comment by HudsonGri** [comment]
I've been prototyping Django model type synthesis for ty and have a working branch I'd like to get direction on before opening PRs: https://github.com/HudsonGri/ruff/tree/django-model-synthesis (~900 lines in `django_model.rs`, 14 mdtest files).

The approach is a Salsa-tracked per-class scanner that reads model bodies and maps field assignments like `title = CharField(max_length=100)` to their Python value types (`str`, `int`, etc.). Instance member lookup gets intercepted in `own_instance_member` so the synthesized type is returned instead of the `Unknown | CharField[str, str]` widening that `place_by_id` would normally produce. Beyond basic fields it handles FK/O2O/M2M relations with reverse accessors, `objects` as `Manager[Model]`, `__init__` synthesis, `pk`/`id`/`_id` accessors, and a few other things; the branch has the full list and known limitations.

I've run it against a production Django codebase and it gets rid of most of the model-related `unresolved-attribute` false positives. I do have an architectural concern with my own approach: the prototype adds `KnownClass` variants for the Django types, which breaks the stdlib-only pattern in that enum. I suspect there's a cleaner way to handle third-party class identity but wasn't sure what the right abstraction would be.

The prototype gates everything behind `[analysis] django = true`; NamedTuple/TypedDict/dataclass synthesis are always-on, so I wasn't sure if this should be too or if opt-in is right for a third-party library. If you're open to upstreaming I'd split into three PRs (field scanning + instance member synthesis; managers + `__init__` + `_id`; reverse relation index).

Pyrefly recently shipped something similar (native, AST-based, no runtime Django dependency).

**Comment by carljm** [comment]
@HudsonGri Very cool, thank you! I took a quick look at it -- the high-level direction looks solid! Would be happy to see the first PR and give it a review.

I think extending `KnownClass` beyond the stdlib is necessary to be able to do built-in support for these third-party libraries;  was expecting we would need to do that. Will take a look at the details in the PR.

My inclination is that this feature should be always-on and not require any additional configuration.

**Comment by HudsonGri** [comment]
Opened the first PR: astral-sh/ruff#24388 with field scanning and instance member synthesis!

**Comment by sobolevn** [comment]
We recently added `ty` support to `django-stubs`: https://github.com/typeddjango/django-stubs/blob/master/ty.toml

However, it does not type check models, of course. Only mypy plugin can do that.

**Comment by ronaldorcampos** [comment]
@HudsonGri thats awesome, I look forward to it!!

**Comment by kalekseev** [comment]
[django-types](https://github.com/sbdchd/django-types/) on main branch works good enought with ty. The only problem I found is ArrayField type, for example `ArrayField(CharField())` results in `list[list[str]]` instead of `list[str]` https://github.com/sbdchd/django-types/blob/661e6c6d6d2d06f5cd65e0288cbaba1a64d6a8dd/tests/trout/models.py#L632

Another problem unrelated to django is that methods shadow types, eg method `list` conflicts with type `list` https://github.com/sbdchd/django-types/blob/661e6c6d6d2d06f5cd65e0288cbaba1a64d6a8dd/django-stubs/contrib/staticfiles/finders.pyi#L10

**Comment by sciyoshi** [comment]
I have a branch with experimental Django support for ty here: https://github.com/sciyoshi/ruff/tree/django-support.

This is not something I’ll be opening as a PR, and it has not been reviewed by me (and frankly I probably wouldn't fully understand what it did given my unfamiliarity with ty internals) in detail. It was built over roughly 14 hours by Codex /goal mode as an exploration of what it would take to make ty usable on a large real-world Django codebase.

What it implements, broadly:

- Django model field descriptor types for common scalar and relational fields
- `ForeignKey`, `OneToOneField`, `ManyToManyField`, and reverse relationship behavior
- model constructor keyword support based on declared fields
- basic manager/queryset typing, including `objects`, `_default_manager`, `_base_manager`, `filter`, `exclude`, `get`, `all`, etc.
- lookup typing for many common queryset filters
- synthesized Django model attributes such as `<field>_id`, reverse accessors, and some generated exception/display helpers
- project-specific model/string reference resolution in common cases

The motivation was practical: mypy is very slow on our codebase, while ty is much faster, so I wanted to see how far ty could get on a large Django project with framework-aware support.

The branch mostly works on that codebase and removes a large number of false positives. That said, it is not fully generic in a few places - some logic uses various heuristics for performance reasons, etc. That made the implementation much cheaper and faster, but it also means it should be treated as an experiment rather than a polished design.

Known limitations include:

- incomplete coverage of Django’s full field and lookup surface
- incomplete support for custom managers/querysets in unusual patterns
- heuristic handling of some queryset/manager naming and inheritance cases
- limited support for highly dynamic model construction or metaclass behavior
- likely missing cases around annotations, custom descriptors, third-party Django extensions, and unusual reverse relationship setup
- structure and code have not been reviewed for quality

_Caveat emptor_ 😅. Still, if anyone is interested in adopting ty on a Django project, this branch may be useful as a starting point for experimentation. In particular, the test cases that were written may be of some value.

**Comment by satya-waylit** [comment]
@sciyoshi mypy 2.0 enables parallel mode that may help. https://mypy-lang.blogspot.com/

**Comment by boxed** [comment]
I'm developing iommi_lsp https://github.com/boxed/iommi_lsp that proxies ty but adds Django specific stuff. There's a lot of very specific stuff I have to do, like handling `__` namespace jumping in Django query expressions. To do that correctly, you have to understand `ForeignKey` mappings, and much more. I'm not sure adding all this to ty itself would make sense?

**Comment by regularkevvv** [comment]
Following up in case this is useful: since [291 was closed as not planned](https://github.com/astral-sh/ty/issues/291), I built `ty-extended`, a fork of `ty` with a generic, sandboxed semantic-extension system:

https://github.com/regularkevvv/ty-extended

Extensions run as WASM in Wasmtime and communicate through versioned requests and declarative patches, without depending on ty’s internal AST, Salsa database, or type objects.

The first live extension is `django-ty`, which has ~100% coverage of the `django-stubs` type support:

https://github.com/regularkevvv/django-ty

Tbh the only reason to exist of the fork is 291 being closed, but I'd be more than happy to open a Pr upstream if the `ty` maintainers change their mind and decide to support plugins. Ty is the best type checker out there and the only Rust based with a friendly license.
